adal/src/main/java/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -849,9 +854,68 @@ private String checkInputParameters(final String resource, final String clientId
             redirectUri = inputRedirectUri;
         }
 
+        //check the permissions required for the broker usage
+        if (AuthenticationSettings.INSTANCE.getUseBroker()) {
+            try {
+                checkBrokerPermissions();
+            } catch (final UsageAuthenticationException exception) {
+                callback.onError(exception);
+            }
+        }
+
         return redirectUri;
     }
 
+    /**
+     * To verify if App gives permissions to AccountManager to use broker. 
+     * If target version is lower than 23, calling app has to have GET_ACCOUNTS, 
+     * MANAGE_ACCOUNTS, USE_CREDENTIALS permissions granted in the Manifest.xml.
+     * Beginning in Android 6.0, the run-time permission GET_ACCOUNTS is required
+     * which need to be requested in the runtime by the calling app.
+     * 
+     * @return true if all required permissions are granted, otherwise the exception will be thrown.
+     * @throws UsageAuthenticationException 
+     */
+    private boolean checkBrokerPermissions() throws UsageAuthenticationException {
+        final String methodName = ":checkBrokerPermissions";
+        final PackageManager pm = mContext.getPackageManager();
+        final StringBuilder permissionMissing = new StringBuilder(46);
+
+        if (pm.checkPermission("android.permission.GET_ACCOUNTS", mContext.getPackageName()) != PackageManager.PERMISSION_GRANTED) {
+            permissionMissing.append("GET_ACCOUNTS ");
```

Comment:
you can have a private method taking int he string builder, and permission name, wrap the append and logging there.